File: densmap/densmap.py
import sys
import numpy as np
import argparse
import pickle

import densmap
from sklearn.datasets import load_digits

def parse_args():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('-i','--input', help='Input .txt or .pkl', default='data.txt')
    parser.add_argument('-o','--outname', help='Output prefix for saving _emb.txt, _dens.txt',
                        default='out')
    parser.add_argument('-f','--dens_frac', type=float, default=0.3)
    parser.add_argument('-l','--dens_lambda', type=float, default=2.0)
    parser.add_argument('-s','--var_shift', type=float, default=0.1)
    parser.add_argument('-d','--ndim', type=int, default=2, help='Embedding dimension (default: %(default)s)')
    parser.add_argument('-n','--n-epochs', type=int, help='Number of epochs', default=750)
    parser.add_argument('-k','--n-nei', type=int, default=30, help='Number of neighbors (default: %(default)s)')
    parser.add_argument('--final_dens', action='store_true', default=True)
    parser.add_argument('--no_final_dens', dest='final_dens', action='store_false')
    parser.add_argument('--outtype', choices=('pkl','txt'), default='txt', help='Output format type (default: %(default)s)')
    return parser

def main(args):
    if args.input.endswith('.txt'):
        data = np.loadtxt(args.input)
    elif args.input.endswith('.pkl'):
        data = pickle.load(open(args.input,'rb'))
    else:
        raise RuntimeError(f'File format for {args.input} not supported')

    if data.shape[0] < data.shape[1]:
        data = data.T
    
    emb = densmap.densMAP(verbose=True,
                          n_components=args.ndim,
                          n_neighbors=args.n_nei,
                          n_epochs=args.n_epochs,
                          dens_frac=args.dens_frac,
                          dens_lambda=args.dens_lambda,
                          logdist_shift=0,
                          var_shift=args.var_shift,
                          final_dens=args.final_dens).fit_transform(data)

    outname = args.outname
    if args.final_dens:
        (emb, ro, re) = emb
        rero = np.stack((ro,re)).transpose()

        if args.outtype=='txt':
            np.savetxt(outname+'_dens.txt',rero, fmt='%e')
        elif args.outtype=='pkl':
            with open(outname + '_dens.pkl','wb') as f:
                pickle.dump(rero, f)
        else:
            raise RuntimeError
        
    if args.outtype == 'txt':
        np.savetxt(outname+'_emb.txt',emb, fmt='%e')

    elif args.outtype == 'pkl':
        with open(outname + '_emb.pkl','wb') as f:
            pickle.dump(emb, f)
    else: # should not reach here
        raise RuntimeError 

    print("Done")

if __name__ == '__main__':
    main(parse_args().parse_args())

File: densmap/densmap.R
if(!require("reticulate")) install.packages("reticulate", dependencies = TRUE, INSTALL_opts = '--no-lock')
# if(!require("optparse")) install.packages("optparse", dependencies = TRUE, INSTALL_opts = '--no-lock')

library("reticulate")
# install.packages("optparse")
library("optparse")

# Installs densMAP via pip
if(reticulate::py_module_available('densmap')) { 
    sys <- reticulate::import('sys')
    sbp <- reticulate::import('subprocess')
    sbp$check_call(c(sys$executable, "-m", "pip", "install", 'densmap-learn'))
} 

densMAP <- function(data, n_neighbors, n_epochs, ndim, dens_frac, dens_lambda, final_dens, var_shift, metric, min_dist,verbose) { 
    if(missing(n_neighbors)) { 
        n_neighbors <- 30
    }
    if(missing(n_epochs)) { 
        n_epochs <- 750
    }
    if(missing(ndim)) { 
        ndim <- 2
    }
    if(missing(dens_frac)) { 
        dens_frac <- .3
    }
    if(missing(dens_lambda)) { 
        dens_lambda <- 2.
    }
    if(missing(final_dens)) { 
        final_dens <- TRUE
    }
    if(missing(var_shift)) { 
        var_shift <- .1
    }
    if(missing(metric)) { 
        metric <- "euclidean"
    }
    if(missing(min_dist)) { 
        min_dist <- .1
    }
    if(missing(verbose)) { 
        verbose <- TRUE
    }
            
    dm <- reticulate::import('densmap')
    
    densMAP <- dm$densMAP("n_neighbors"=as.integer(n_neighbors),"n_epochs"=as.integer(n_epochs), 
                          "n_components"=as.integer(ndim), "dens_frac"=dens_frac, "dens_lambda"=dens_lambda, 
                          "final_dens"=final_dens, "var_shift"=var_shift, "metric"=metric,
                          "min_dist"=min_dist, "verbose"=verbose)
    
    mdm.full <- densMAP$fit_transform(data)
        
    return(mdm.full)
}

#dm <- reticulate::import('densmap')

#data <- dm$load_trial$load_trial()
#densMAP <- dm$densMAP("final_dens"=FALSE)

#mdm.full <- densMAP$fit_transform(data)

#mdm.full